nodeletctl/pkg/nodeletctl/certs.go

```diff
@@ -234,3 +234,47 @@ func writeKubeconfigFile(args *KubeConfigData) error {
 	zap.S().Infof("Wrote kubeconfig to %s\n", kubeconfigFile)
 	return nil
 }
+
+func RenewCAIfExpiring(cfg *BootstrapConfig) error {
+	certsDir := filepath.Join(ClusterStateDir, cfg.ClusterId, "certs")
+	caCertFile := filepath.Join(certsDir, RootCACRT)
+
+	caFile, err := ioutil.ReadFile(caCertFile)
+	if err != nil {
+		return fmt.Errorf("Failed to read CA Cert %s: %s", caCertFile, err)
+	}
+	caPEM, _ := pem.Decode(caFile)
+	ca, err := x509.ParseCertificate(caPEM.Bytes)
+	if err != nil {
+		return fmt.Errorf("Failed to parse CA Cert %s: %s", caCertFile, err)
+	}
+
+	currTime := time.Now()
+	expireTime := ca.NotAfter
+
+	diffTime := expireTime.Sub(currTime)
+	daysTillExpiry := int64(diffTime.Hours() / 24)
+	if daysTillExpiry < CAExpiryThreshold {
+		zap.S().Infof("Cert is expiring in %d days (%d hours), will re-generate", daysTillExpiry, diffTime.Hours())
+		return RegenCA(cfg)
+	}
+	return nil
+}
+
+func RegenCA(cfg *BootstrapConfig) error {
+	certsDir := filepath.Join(ClusterStateDir, cfg.ClusterId, "certs")
+	if err := os.RemoveAll(certsDir); err != nil {
+		return fmt.Errorf("Failed to remove old certs directory: %s", err)
+	}
+
+	_, err := GenCALocal(cfg.ClusterId)
```

Comment:
I thought the idea was to re-use the original key when regenerating CA. `GenCALocal` calls `GenCA` which creates a new random key. I might have missed some communication in case it was already discussed.

Reply:
Thats what I thought too, but we have the CA priv key on each node. Roopak said we should change the key as well.